plugin/trino-hive/src/main/java/io/trino/plugin/hive/parquet/ParquetPageSourceFactory.java

```diff
@@ -238,37 +274,60 @@ public static ReaderPageSource createPageSource(
+            }
 
             long nextStart = 0;
             ImmutableList.Builder<BlockMetaData> blocks = ImmutableList.builder();
             ImmutableList.Builder<Long> blockStarts = ImmutableList.builder();
             ImmutableList.Builder<Optional<ColumnIndexStore>> columnIndexes = ImmutableList.builder();
             for (BlockMetaData block : parquetMetadata.getBlocks()) {
                 long firstDataPage = block.getColumns().get(0).getFirstDataPageOffset();
-                Optional<ColumnIndexStore> columnIndex = getColumnIndexStore(dataSource, block, descriptorsByPath, parquetTupleDomain, options);
-                Optional<BloomFilterStore> bloomFilterStore = getBloomFilterStore(dataSource, block, parquetTupleDomain, options);
-
-                if (start <= firstDataPage && firstDataPage < start + length
-                        && predicateMatches(
-                        parquetPredicate,
-                        block,
-                        dataSource,
-                        descriptorsByPath,
-                        parquetTupleDomain,
-                        columnIndex,
-                        bloomFilterStore,
-                        timeZone,
-                        domainCompactionThreshold)) {
-                    blocks.add(block);
-                    blockStarts.add(nextStart);
-                    columnIndexes.add(columnIndex);
+                boolean isBlockAlreadyAdded = false;
+                for (ParquetDomainAndPredicate domain : parquetDomains) {
+                    if (isBlockAlreadyAdded) {
+                        break;
+                    }
+
+                    TupleDomain<ColumnDescriptor> parquetTupleDomain = domain.tupleDomain;
+                    TupleDomainParquetPredicate parquetPredicate = domain.predicate;
+                    Optional<ColumnIndexStore> columnIndex = getColumnIndexStore(dataSource, block, descriptorsByPath, parquetTupleDomain, options);
+                    Optional<BloomFilterStore> bloomFilterStore = getBloomFilterStore(dataSource, block, parquetTupleDomain, options);
+                    if (start <= firstDataPage && firstDataPage < start + length
+                            && predicateMatches(
+                                    parquetPredicate,
+                                    block,
+                                    dataSource,
+                                    descriptorsByPath,
+                                    parquetTupleDomain,
+                                    columnIndex,
+                                    bloomFilterStore,
+                                    timeZone,
+                                    domainCompactionThreshold)) {
+                        blocks.add(block);
+                        blockStarts.add(nextStart);
+                        columnIndexes.add(columnIndex);
+                        isBlockAlreadyAdded = true;
+                    }
+                    nextStart += block.getRowCount();
```

Comment:
Is it correct that we increment this for every `parquetDomain`?

Reply:
Good catch.